pkg/sharding/apiserver/handler.go

```diff
@@ -0,0 +1,359 @@
+		Resource:                    resource,
+		Subresource:                 requestInfo.Subresource,
+		MetaGroupVersion:            metav1.SchemeGroupVersion,
+		HubGroupVersion:             groupVersion,
+		MaxRequestBodyBytes:         h.maxRequestBodyBytes,
+		AcceptsGroupVersionDelegate: everythingAcceptor{},
+	}
+
+	nopAdmission := admission.NewHandler()
+
+	userInfo, ok := request.UserFrom(req.Context())
+	if !ok {
+		responsewriters.ErrorNegotiated(
+			apierrors.NewInternalError(fmt.Errorf("missing userInfo")),
+			Codecs, schema.GroupVersion{Group: requestInfo.APIGroup, Version: requestInfo.APIVersion}, w, req,
+		)
+		return
+	}
+
+	s := &shardedStorage{
+		TableConvertor:                 tableConverter,
+		shards:                         h.clients,
+		shardIdentifierResourceVersion: 1,
+		requestFor:                     requestFor(req, requestInfo),
+		clientFor:                      clientFor(userInfo, negotiatedSerializer),
+	}
+
+	var handlerFunc http.HandlerFunc
+	switch requestInfo.Verb {
+	case "get":
+		handlerFunc = handlers.GetResource(s, requestScope)
+	case "list":
+		forceWatch := false
+		handlerFunc = handlers.ListResource(s, s, requestScope, forceWatch, h.requestTimeout)
+	case "watch":
+		forceWatch := true
+		handlerFunc = handlers.ListResource(s, s, requestScope, forceWatch, h.requestTimeout)
+	case "create":
+		handlerFunc = handlers.CreateResource(s, requestScope, nopAdmission)
+	case "update":
+		handlerFunc = handlers.UpdateResource(s, requestScope, nopAdmission)
+	case "patch":
+		supportedTypes := []string{
+			string(types.JSONPatchType),
+			string(types.MergePatchType),
+		}
+		if legacyscheme.Scheme.IsGroupRegistered(requestInfo.APIGroup) {
+			supportedTypes = append(supportedTypes, string(types.StrategicMergePatchType))
+		}
+		if utilfeature.DefaultFeatureGate.Enabled(features.ServerSideApply) {
+			supportedTypes = append(supportedTypes, string(types.ApplyPatchType))
+		}
+		handlerFunc = handlers.PatchResource(s, requestScope, nopAdmission, supportedTypes)
+	case "delete":
+		allowsOptions := true
+		handlerFunc = handlers.DeleteResource(s, allowsOptions, requestScope, nopAdmission)
+	case "deletecollection":
+		checkBody := true
+		handlerFunc = handlers.DeleteCollection(s, checkBody, requestScope, nopAdmission)
+	default:
+		responsewriters.ErrorNegotiated(
+			apierrors.NewMethodNotSupported(schema.GroupResource{Group: requestInfo.APIGroup, Resource: requestInfo.Resource}, requestInfo.Verb),
+			Codecs, schema.GroupVersion{Group: requestInfo.APIGroup, Version: requestInfo.APIVersion}, w, req,
+		)
+		handlerFunc = nil
+	}
+	if handlerFunc != nil {
+		handlerFunc = metrics.InstrumentHandlerFunc(requestInfo.Verb, requestInfo.APIGroup, requestInfo.APIVersion, requestInfo.Resource, requestInfo.Subresource, metrics.CleanScope(requestInfo), metrics.APIServerComponent, false, "", handlerFunc)
+		handler := genericfilters.WithWaitGroup(handlerFunc, longRunningFilter, h.wg)
```

Comment:
for what is that needed? It's something to synchronize audit with the request chain. I don't think you care here.